cli/WFCORE-2663_CLI_match open_close_characters.adoc

```diff
@@ -0,0 +1,61 @@
+= WFCORE-2663 Match open/close characters on CLI commands
+:author:            Ingo Weiss
+:email:             ingo@redhat.com
+:toc:               left
+:icons:             font
+:keywords:          comma,separated,tags
+:idprefix:
+:idseparator:       -
+:issue-base-url:    https://issues.jboss.org/browse/
+
+== Overview
+
+When using the CLI to configure WildFly, sometimes commands with multiple options and parameters require the use of `( )`, `[ ]` and `{ }`. This becomes hard to read once there are line breaks for the same command. Matching the opening and closing characters of these pairs help legibility
+
+In a context where we have a nesting of complex objects as operation parameters, it is difficult to match `(`, `{` and `[` with their closing counterparts. In particular at the end of the command, it is needed to quickly identify missing closing characters.
+
+We are proposing here to highlight the matching pairs when the cursor is on the closing or opening character.
+
+== Issue Metadata
+
+=== Issue:
+
+* {issue-base-url}WFCORE-2663[WFCORE-2663]
+
+=== Related issues:
+
+* {issue-base-url}EAP-737[EAP-737]
+
+=== Dev Contacts:
+
+* mailto:ingo@redhat.com[Ingo Weiss]
+* mailto:jdenise@redhat.com[Jean-François Denise]
+
+=== QE Contacts:
+
+* mailto:tterem@redhat.com[Tomas Terem]
+
+=== Affected Projects or Components:
+
+* WildFly Core CLI
+
+== Requirements
+
+=== Hard requirements
+
+* When moving the cursor onto an opening character ([, {, (), the closing character (], }, )) should be highlighted in green. If there are unmatched brackets between the matched pair, then these brackets will be highlighted in red;
```

Comment:
We need to specify how to match an opening character with its closing counterpart and vice versa. It seemed to be intuitive, but during implementation and testing, it turned out that it is not.
E.g. which one is the opening bracket for bold bracket here:
( { [ ( ] } **)**
( ( **)**
( ( **)** )
Is it the first or second '('? Or is the answer different for different situations?
@iweiss, as we discussed, both approaches seems to be counter-intuitive in different situations. We need to either choose one of them or combine them together. Chosen approach needs to be described in the analysis.
cc: @jfdenise

Reply:
The closing character is closing the nearest opening one. I see it as a stack of opening characters, when an opening character is found, it is pushed to the stack. When a closing character is found, the last character in the stack is un-stacked and matched.

Reply:
@jfdenise This is how it's working right now, https://asciinema.org/a/aiOqbBolpy2x7ZWcyhh3HZqVG. As you can see, since we don't mark anything out of the scope and we choose the closest "closing" character, the middle `(` ends up not being marked.

If we change the behaviour to do a full analysis of the line, then we see the "closing" character changing position based on where the cursor is and what's being taken in consideration.

I think if we could have access to the CLI parser this wouldn't be a problem. What do you think?

Reply:
What I saw in the recording is working well, what is outside the scope the first opening ( is not marked in red, that seems fine to me. I feel to not follow the issue there.

Reply:
@tterem I don't see ambiguity in your use cases:
( { [ ( ] } **)** ==> The 4th character is the opening one.
( ( **)** ==> The 2nd character is the opening one.
( ( **)** ) ==> The 2nd character is the opening one.
Makes sense or you see other cases?

Reply:
@jfdenise 

> The closing character is closing the nearest opening one. I see it as a stack of opening characters, when an opening character is found, it is pushed to the stack. When a closing character is found, the last character in the stack is un-stacked and matched.
> ( { [ ( ] } ) ==> The 4th character is the opening one.
( ( ) ==> The 2nd character is the opening one.
( ( ) ) ==> The 2nd character is the opening one.

Okay, I agree with this approach. @iweiss can you put it to the analysis document?